### PR TITLE
bootstrap: allow building with shared BlocksRuntime

### PR DESCRIPTION
**Pull request by compnerd**

### CODE REVIEW

File: Utilities/bootstrap [hunk 1]
```diff
@@ -579,6 +579,8 @@ class llbuild(object):
                 if self.args.foundation_path:
                     link_command.extend(["-L", self.args.foundation_path])
                 if self.args.libdispatch_build_dir:
+                    link_command.extend(['-L', self.args.libdispatch_build_dir,
+                                         '-Xlinker', '-lBlocksRuntime'])
```
Comment (lanza) [inline Utilities/bootstrap]: Shouldn't this be
     `link_command.extend(['-L', self.args.libdispatch_build_dir, '-lBlocksRuntime'])`

Comment (compnerd) [reply Utilities/bootstrap]: It seemed that the actual linked library names are passed directly through the driver in the rest of the build.  Im just following the pattern.  I believe that the driver will forward the options either way.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
The failure is weird, llbuild linkage is failing, which should not be impacted by addition only linkage for BlocksRuntime.

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@aciidb0mb3r - okay to merge this?

**Comment by compnerd** [comment]
@swift-ci please smoke test

File: Utilities/bootstrap [hunk 2]
```diff
@@ -1125,6 +1127,9 @@ def main():
             symlink_force(os.path.join(args.foundation_path, 'libFoundation.so'),
                 libswiftdir)
             if args.libdispatch_build_dir:
+                symlink_force(os.path.join(args.libdispatch_build_dir,
+                                           'libBlocksRuntime.so'),
+                              libswiftdir)
```
Comment (aciidb0mb3r) [inline Utilities/bootstrap]: Why do we need this symlink?

**Comment by aciidb0mb3r** [comment]
Can you explain what is this and why do clients need to link to it manually? Also, is this a hard dependency?

File: Utilities/bootstrap (hunk 2 above)
Comment (compnerd) [reply Utilities/bootstrap]: Without that symlink, the tests tended to fail.  libdispatch would not be found, but by keeping the symlink there, which is one of the paths that linker/loader consult, the tests are happier.

**Comment by compnerd** [comment]
libBlocksRuntime is the runtime for blocks.  The closures in swift are actually blocks, and need a runtime for to support them.  Currently, it was built static and put into libdispatch and a couple of other entities.  However, there are globals which need to be uniqued across the image, and that requires that we have the BlocksRuntime be shared.  This is setup to convert the build of the Blocks Runtime to a shared library.

**Comment by aciidb0mb3r** [comment]
Thanks for the explanation, @compnerd! It sounds like this is a dependency of the libdispatch or the Swift compiler itself. I don't think SwiftPM (or every Swift based Swift.org project) should need to explicitly configure search paths for this dependency.

**Comment by compnerd** [comment]
@aciidb0mb3r, unfortunately, during the complex CI build, that isnt really feasible.  When the components are installed in the proper location (`/usr/lib`), they will be found without a search path.  Switching over to the shared build will allow us to move the library into swift itself, and the search path can be removed at that point.  I think it would be more reasonable to try to move this into the driver at that point.  This change is basically staging work.

**Comment by aciidb0mb3r** [comment]
@compnerd Ah, okay. That sounds reasonable to me then! Feel free to merge.

**Pull request merged**